Sprint 07

shooin13 opened the pull request:
- MVP refactoring
- Unit tests
- UI Tests

shooin13 committed:
Added UI tests

shooin13 committed:
Revert "Added UI tests"

This reverts commit 79e46aceff99f410d5ba89322313d0d0d1fa7ff1.

shooin13 committed:
Revert "Revert "Added UI tests""

This reverts commit 8250670653b69d347c7c7682d64c537657a99f42.

shooin13 committed:
Began MVP refactoring: moved 'convert' method out of MovieQuizController to MovieQuizPresenter

shooin13 committed:
Moved buttons tap logic to MovieQuizPresenter

shooin13 committed:
[continue MVP refactoring] Moved buttons functionality and didReceiveNextQuestion, showNextQuestionOrResult methods to presenter

shooin13 committed:
[continue MVP refactoring] normalization according DRY: moved correctAnswers and questionFactory to MovieQuizPresenter

shooin13 committed:
[continue MVP refactoring] moved statisticService to MovieQuizPresenter
- renamed showAnswerResult to proceedWithAnswer
- renamed showNextQuestionOrResults to proceedToNextQuestionOrResults
- cleaned code (applied correct MARKs)

shooin13 committed:
Added MovieQuizPresenter tests

shooin13 committed:
[Finished MVP refactoring], cleaned code

shooin13 committed:
Added testing of closing quiz result alert and resetting quiz indexLabel
Deleted not needed test

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,172 @@
+  private var alertPresenter: AlertPresenter?
+  private weak var viewController: MovieQuizViewController?
+  private var questionFactory: QuestionFactoryProtocol?
+  private var statisticService = StatisticServiceImplementation()
```
После рефакторинга, потерял протокол под который должна быть подписана данная переменная. Я ожидаю как в предыдущих спринтах увидеть:
```
private var statisticService: StatisticService?
```
А ниже в init()
```
statisticService = StatisticServiceImplementation()
```

shooin13 replied:
Привет! Спасибо, поправил!

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,172 @@
+  // MARK: - Initialization
+  
+  init (viewController: MovieQuizViewControllerProtocol) {
+    self.viewController = viewController as? MovieQuizViewController
```
Нет необходимости кастить `viewController` до `MovieQuizViewController` он уже сам по себе соответствует этому протоколу, также на 11-ой строке необходимо указать, что он соответствует протоколу `private weak var viewController: MovieQuizViewControllerProtocol?`
Иначе твой подход нарушает архитектуру MVP

После этого будут ошибки связанные с тем, что свойство `weak var viewController` под протоколом `MovieQuizViewControllerProtocol` не знает о наличие например `buttonsStackView`. Для этого тебе нужно в протоколе реализовать методы, которые взаимодействуют с твоим `stackView`.

Это все поможет корректно реализовать архитектуру MVP, без нарушений

shooin13 replied:
Сделано, спасибо!

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,172 @@
+  
+  // MARK: - Button Actions
+  
+  func yesButtonClicked() {
```
Этот метод и следующие после него не используются вне класса, давай используем модификатор `private` и инкапсулируем все, что не должно торчать наружу, внутри Presenter-а

shooin13 replied:
методы yesButtonClicked() и noButtonClicked() используются в MovieQuizViewController

Остальные, которые не используются  - проверил, сделал приватными те, которые не используются

So1ovyov replied:
Все верно методы `yesButtonClicked` и `noButtonClicked` как раз используются вне презентера, у них `private` не нужен 👍🏻

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,197 +1,119 @@
-  private let moviesLoader = MoviesLoader()
-  
-
+  private var presenter: MovieQuizPresenter!
```
Хорошо бы обезопасить проект и использовать не `force property`, а опционал `private var presenter: MovieQuizPresenter?`

shooin13 replied:
Принято!

Так реализовано в теории, в ЯП. Меняю, спасибо!

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,197 +1,119 @@
+                                accessibilityIndicator: "QuizResultAlert") { [weak self] in
       guard let self else { return }
-      alertPresenterDidTapButton(alertPresenter)
+      presenter.alertPresenterDidTapButton(restart: true)
```
В этом методе и ниже после ослабления `self`, в замыкании ты используешь `presenter`, который не ослабляешь, из-за чего возможна утечка памяти.

shooin13 replied:
Если честно, я не до конца понял эту ошибку, но поменял, как ты порекомендовал. Буду признателен, если расскажешь чуть подробнее.

So1ovyov replied:
Вот что я примерно ожидал увидеть, ослабить `self` в клоужере
```
    func showNetworkError(message: String) {
        hideLoadingIndicator()
        presenter?.showAlert(
            .init(
                title: "Что-то пошло не так(",
                message: message,
                buttonText: "Попробовать еще раз",
                accessibilityIndicator: "NetworkErrorAlert"
            ) { [weak self] in
                guard let self else { return }
                self.presenter?.loadQuestionData()
            }
        )
    }
```
А вот неплохая статья как перестраховаться от утечек памяти https://habr.com/ru/companies/otus/articles/666436/

shooin13 replied:
Большое спасибо! Прочитаю статью, потом вернусь к твоему ожиданию и дополню.

So1ovyov commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,21 @@
+import XCTest
+@testable import MovieQuiz
+
+class ArrayTests: XCTestCase {
```
Давай добавим модификатор `final` для этого класса. Он необходим для оптимизации и скорости проекта, также не позволит наследоваться от данного класса.

shooin13 replied:
Здесь глупо вышло, поставил final, благодарю!

So1ovyov commented on MovieQuizUITests/MovieQuizUITestsLaunchTests.swift:
```diff
@@ -0,0 +1,25 @@
+import XCTest
+
+final class MovieQuizUITestsLaunchTests: XCTestCase {
```
Данный дефолтный тест сгенерирован xcode и его можно удалить.

shooin13 replied:
Принял, удалил)

shooin13 commented:
@So1ovyov Большое спасибо за комментарии!
Пришлось добавить каст в AlertPresenter 
let viewController = delegate?.viewControllerForAlertPresenting() as? MovieQuizViewController, так как медод showAlert содержит present, и MovieQuizViewControllerProtocol' has no member 'present'

shooin13 committed:
Changes according to reviewer comments:
- statisticService type changed from StatisticServiceImplementation() to StatisticService?
- incapsulated methods which are not used in others classes
- changed force property presenter: MovieQuizPresenter! to presenter: MovieQuizPresenter?
- removed casting self.viewController = viewController as? MovieQuizViewController in MovieQuizPresenter initialization, added missing methods to MovieQuizViewControllerProtocol
- got rid of retain cycle in MovieQuizViewController
- added final to class ArrayTests
- removed unused default test

shooin13 committed:
removed commented not needed code

So1ovyov commented:
> @So1ovyov Большое спасибо за комментарии! Пришлось добавить каст в AlertPresenter let viewController = delegate?.viewControllerForAlertPresenting() as? MovieQuizViewController, так как медод showAlert содержит present, и MovieQuizViewControllerProtocol' has no member 'present'

Каст тут не нужен, `present` содержит сам `ViewController` и для того, чтобы вызвать этот метод надо было добавить метод в протоколы и провести их по цепочке MVP
Вот, что ты можешь сделать

1. Добавляем метод  в делегат
```

protocol AlertPresenterDelegate: AnyObject {
    func presentAlert(_ alert: UIAlertController, animated: Bool)

```

2. Вызываем делегат в `AlertPresenter` в методе `showAlert`
```
    func showAlert(_ alertModel: AlertModel) {
        let alert = UIAlertController(
            title: alertModel.title,
            message: alertModel.message,
            preferredStyle: .alert
        )
        alert.view.accessibilityIdentifier = alertModel.accessibilityIndicator

        let action = UIAlertAction(
            title: alertModel.buttonText,
            style: .default
        ) { _ in
            alertModel.completion?()
        }

        alert.addAction(action)
        delegate?.presentAlert(alert, animated: true)
    }
```

3. Используем новый метод делегата в `MoviesQuizPresenter`
```
  //MARK: - AlertPresenterDelegate

    func presentAlert(_ alert: UIAlertController, animated: Bool) {
        viewController?.present(alert: alert, animated: animated)
    }
```

4. Добавляем новый метод в протокол `MovieQuizViewControllerProtocol` чтобы в презентере этот метод был и без каста
```
protocol MovieQuizViewControllerProtocol: AnyObject {
    func present(alert: UIAlertController, animated: Bool)
```

5. Реализовываем новый метод протокола ViewController-а 
```
final class MovieQuizViewController: UIViewController, MovieQuizViewControllerProtocol {

    func present(alert: UIAlertController, animated: Bool) {
        present(alert, animated: animated)
    }
```
